### ISSUE TITLE
[identify/delta] protocol change has race condition

### ISSUE THREAD

**Issue by lanzafame** [libp2p/go-libp2p#686]
This race causes a panic in the wild, see ipfs/ipfs-cluster#845.

Here is the output of the `TestIdentifyDeltaOnProtocolChange` test with `-race` set:
```
=== RUN   TestIdentifyDeltaOnProtocolChange
==================
WARNING: DATA RACE
Read at 0x00c0002d4330 by goroutine 42:
  runtime.mapiterinit()
      /usr/lib/go/src/runtime/map.go:804 +0x0
  github.com/libp2p/go-libp2p-peerstore/pstoremem.(*memoryProtoBook).GetProtocols()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-peerstore@v0.1.1/pstoremem/protobook.go:108 +0x210
  github.com/libp2p/go-libp2p-peerstore.(*peerstore).GetProtocols()
      <autogenerated>:1 +0x88
  github.com/libp2p/go-libp2p/p2p/protocol/identify_test.TestIdentifyDeltaOnProtocolChange()
      /home/lanzafame/src/github.com/libp2p/go-libp2p/p2p/protocol/identify/id_test.go:263 +0xdc8
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:865 +0x163

Previous write at 0x00c0002d4330 by goroutine 76:
  runtime.mapdelete_faststr()
      /usr/lib/go/src/runtime/map_faststr.go:297 +0x0
  github.com/libp2p/go-libp2p-peerstore/pstoremem.(*memoryProtoBook).RemoveProtocols()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-peerstore@v0.1.1/pstoremem/protobook.go:127 +0x1ca
  github.com/libp2p/go-libp2p-peerstore.(*peerstore).RemoveProtocols()
      <autogenerated>:1 +0xa6
  github.com/libp2p/go-libp2p/p2p/protocol/identify.(*IDService).consumeDelta()
      /home/lanzafame/src/github.com/libp2p/go-libp2p/p2p/protocol/identify/id_delta.go:71 +0x155
  github.com/libp2p/go-libp2p/p2p/protocol/identify.(*IDService).deltaHandler()
      /home/lanzafame/src/github.com/libp2p/go-libp2p/p2p/protocol/identify/id_delta.go:37 +0x5af
  github.com/libp2p/go-libp2p/p2p/protocol/identify.(*IDService).deltaHandler-fm()
      /home/lanzafame/src/github.com/libp2p/go-libp2p/p2p/protocol/identify/id_delta.go:17 +0x55
  github.com/libp2p/go-libp2p-blankhost.(*BlankHost).SetStreamHandler.func1()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-blankhost@v0.1.3/blank.go:125 +0xb4

Goroutine 42 (running) created at:
  testing.(*T).Run()
      /usr/lib/go/src/testing/testing.go:916 +0x65a
  testing.runTests.func1()
      /usr/lib/go/src/testing/testing.go:1157 +0xa8
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:865 +0x163
  testing.runTests()
      /usr/lib/go/src/testing/testing.go:1155 +0x523
  testing.(*M).Run()
      /usr/lib/go/src/testing/testing.go:1072 +0x2eb
  main.main()
      _testmain.go:54 +0x222

Goroutine 76 (finished) created at:
  github.com/libp2p/go-libp2p-blankhost.(*BlankHost).newStreamHandler()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-blankhost@v0.1.3/blank.go:156 +0x2bf
  github.com/libp2p/go-libp2p-blankhost.(*BlankHost).newStreamHandler-fm()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-blankhost@v0.1.3/blank.go:146 +0x55
  github.com/libp2p/go-libp2p-swarm.(*Conn).start.func1.1()
      /home/lanzafame/pkg/mod/github.com/libp2p/go-libp2p-swarm@v0.1.0/swarm_conn.go:116 +0x120
==================
--- FAIL: TestIdentifyDeltaOnProtocolChange (1.30s)
    testing.go:809: race detected during execution of test
```

**Comment by raulk** [comment]
This was already fixed and released in go-libp2p-peerstore v0.1.2; however, we had not released go-libp2p after that. I've just done that now. Please upgrade to go-libp2p v0.2.1: https://github.com/libp2p/go-libp2p/releases/tag/v0.2.1.

**Comment by lanzafame** [comment]
@raulk thanks!
